---
title: "RNAseq_DESeq2"
output: html_notebook
---

First time organising R as a project/with github. Referring to this walkthrough on the topic:
https://ourcodingclub.github.io/2017/02/27/git.html

sampleTables
```{r}
sampleTableExpt1 <- read.csv("~/rnaSeqData/sampleTables/sampleTableExpt1.csv")
sampleTableExpt2 <- read.csv("~/rnaSeqData/sampleTables/sampleTableExpt2.csv")
sampleTableExpt3 <- read.csv("~/rnaSeqData/sampleTables/sampleTableExpt3.csv")
```

tximport
```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("tximport")
library("tximport")

myDir <- "/home/mcity599/rnaSeqData/salmonAlignments/"
expt1Samples <- c("cag2430_1GCBias","cag2430_2GCBias","cag2430_3GCBias","cag2430_4GCBias","cag2430_5GCBias","cag2430_6GCBias")
expt1Files <- file.path(myDir,expt1Samples,"quant.sf")
names(expt1Files) <- paste0("cag2430_",1:6)
all(file.exists(expt1Files))
expt1Files

myDir <- "/home/mcity599/rnaSeqData/salmonAlignments/"
expt2Samples <- c("cag2430_7GCBias","cag2430_8GCBias","cag2430_9GCBias","cag2430_10GCBias","cag2430_11GCBias","cag2430_12GCBias")
expt2Files <- file.path(myDir,expt2Samples,"quant.sf")
names(expt2Files) <- paste0("cag2430_",7:12)
all(file.exists(expt2Files))
expt2Files

myDir <- "/home/mcity599/rnaSeqData/salmonAlignments/"
expt3Samples <- c("cag2430_13GCBias","cag2430_14GCBias","cag2430_15GCBias","cag2430_16GCBias","cag2430_17GCBias","cag2430_18GCBias")
expt3Files <- file.path(myDir,expt3Samples,"quant.sf")
names(expt3Files) <- paste0("cag2430_",13:18)
all(file.exists(expt3Files))
expt3Files


## Create a tx2gene table:
# From Xenbase I've got this:
NcbiMrnaXenbaseGene_laevis <- read.delim("~/rnaSeqData/geneConverters/NcbiMrnaXenbaseGene_laevis.txt", header=FALSE)
dim(NcbiMrnaXenbaseGene_laevis)
# The required format (for tximport) is two columns, transcript and geneID
NcbiMrnaXenbaseGene_laevis[1:5,]
transcriptToGene <- NcbiMrnaXenbaseGene_laevis[,c(2,4)]
colnames(transcriptToGene) <- c("TXNAME","GENEID") # copying the tximport vignette names

expt1TXI <- tximport(expt1Files,type = "salmon", tx2gene = transcriptToGene)
expt2TXI <- tximport(expt2Files,type = "salmon", tx2gene = transcriptToGene)
expt3TXI <- tximport(expt3Files,type = "salmon", tx2gene = transcriptToGene)
```

DESeq2
```{r}
biocLite("DESeq2")
library("DESeq2")

expt1ddsTxi <- DESeqDataSetFromTximport(expt1TXI,
                                   colData = sampleTableExpt1,
                                   design = ~ condition)
expt2ddsTxi <- DESeqDataSetFromTximport(expt2TXI,
                                   colData = sampleTableExpt2,
                                   design = ~ condition)
expt3ddsTxi <- DESeqDataSetFromTximport(expt3TXI,
                                   colData = sampleTableExpt3,
                                   design = ~ condition)
## Edit on 20180219
# While re-reading the tximport vignette I found this important piece of info:
## Note: there are two suggested ways of importing estimates for use with differential gene expression (DGE) methods. The first method, which we show below for edgeR and for DESeq2, is to use the gene-level estimated counts from the quantification tools, and additionally to use the transcript-level abundance estimates to calculate a gene-level offset that corrects for changes to the average transcript length across samples. The code examples below accomplish these steps for you, keeping track of appropriate matrices and calculating these offsets. For edgeR you need to assign a matrix to y$offset, but the function DESeqDataSetFromTximport takes care of creation of the offset for you. Let’s call this method “original counts and offset”.
# The second method is to use the tximport argument countsFromAbundance="lengthScaledTPM" or "scaledTPM", and then to use the gene-level count matrix txi$counts directly as you would a regular count matrix with these software. Let’s call this method “bias corrected counts without an offset”
# Do not do this: Do not manually pass the original gene-level counts to downstream methods without an offset. The original gene-level counts are in txi$counts when tximport was run with countsFromAbundance="no". This is simply passing the summed estimated transcript counts, and does not correct for potential differential isoform usage (the offset), which is the point of the tximport methods (Soneson, Love, and Robinson 2015) for gene-level analysis. Passing uncorrected gene-level counts without an offset is not recommended by the tximport package authors. The two methods we provide here are: “original counts and offset” or “bias corrected counts without an offset”. Passing txi to DESeqDataSetFromTximport as outlined below is correct: the function creates the appropriate offset for you to perform gene-level differential expression.
## I hadn't noticed this when I first did it, but fortunately I had used the correct function anyway (due to the tximport vignette being well written)

# Pre-filtering ( copied from the DESeq2 walkthrough)
# While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful: by removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within DESeq2. Here we perform a minimal pre-filtering to keep only rows that have at least 10 reads total. Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.
keep <- rowSums(counts(expt1ddsTxi)) >= 10
ddsExpt1 <- expt1ddsTxi[keep,]
keep <- rowSums(counts(expt2ddsTxi)) >= 10
ddsExpt2 <- expt2ddsTxi[keep,]
keep <- rowSums(counts(expt3ddsTxi)) >= 10
ddsExpt3 <- expt3ddsTxi[keep,]
rm(keep)

ddsExpt1$condition

ddsExpt1 <- DESeq(ddsExpt1)
ddsExpt2 <- DESeq(ddsExpt2)
ddsExpt3 <- DESeq(ddsExpt3)
# For the sake of size/tidy workspace, I'll remove the objects needed to make ddsExpt 
rm(expt1ddsTxi)
rm(expt2ddsTxi)
rm(expt3ddsTxi)
rm(expt1TXI)
rm(expt2TXI)
rm(expt3TXI)

# LFC shrinkage for transcripts with low read counts:
 # Full methods are described in the DESeq2 paper (see DESeq2 citation), but in short, it looks at the largest fold changes that are not due to low counts and uses these to inform a prior distribution. So the large fold changes from genes with lots of statistical information are not shrunk, while the imprecise fold changes are shrunk. This allows you to compare all estimated LFC across experiments, for example, which is not really feasible without the use of a prior.
#  - https://support.bioconductor.org/p/77461/
resultsNames(ddsExpt1)
resLFCexpt1 <- lfcShrink(ddsExpt1, coef=2)
resLFCexpt2 <- lfcShrink(ddsExpt2, coef=2)
resLFCexpt3 <- lfcShrink(ddsExpt3, coef=2)

rm(ddsExpt1)
rm(ddsExpt2)
rm(ddsExpt3)

# Trim to include only statistically significant results:
statSigExpt1 <- subset(resLFCexpt1, padj < 0.05)
statSigExpt2 <- subset(resLFCexpt2, padj < 0.05)
statSigExpt3 <- subset(resLFCexpt3, padj < 0.05)

# Pull out only genes with biologically significant (i.e., log2FC >1 or < - 1) differences:
Expt1_upregInB <- rownames(statSigExpt1[order(statSigExpt1$log2FoldChange, decreasing = TRUE)[1:sum(statSigExpt1$log2FoldChange > 1, na.rm=T)], ])
Expt1_downregInB <- rownames(statSigExpt1[order(statSigExpt1$log2FoldChange, decreasing = F)[1:sum(statSigExpt1$log2FoldChange < -1, na.rm=T)], ])

Expt2_upregInB <- rownames(statSigExpt2[order(statSigExpt2$log2FoldChange, decreasing = TRUE)[1:sum(statSigExpt2$log2FoldChange > 1, na.rm=T)], ])
Expt2_downregInB <- rownames(statSigExpt2[order(statSigExpt2$log2FoldChange, decreasing = F)[1:sum(statSigExpt2$log2FoldChange < -1, na.rm=T)], ])

Expt3_upregInB <- rownames(statSigExpt3[order(statSigExpt3$log2FoldChange, decreasing = TRUE)[1:sum(statSigExpt3$log2FoldChange > 1, na.rm=T)], ])
Expt3_downregInB <- rownames(statSigExpt3[order(statSigExpt3$log2FoldChange, decreasing = F)[1:sum(statSigExpt3$log2FoldChange < -1, na.rm=T)], ])

## Edit 20180313
# After spending some time with PANTHER, it's clear the Expt1 lists are too small to generate anything significant. I will now generate a less stringent set of results for the purpose of building a picture of what is happening in the WT vs NR. This will be used to inform pathway building and as background gene lists (e.g., as a way to semi-exclude genes that appear to be involved in regen but are upreg in NR vs WT).
# The lowStrin versions will be STATISTICALLY significant, but will have relaxed biological significance (i.e., )
Expt1lowStrin_OverExpressedNR <- rownames(statSigExpt1[order(statSigExpt1$log2FoldChange, decreasing = TRUE)[1:sum(statSigExpt1$log2FoldChange > 0.5, na.rm=T)], ])
Expt1lowStrin_OverExpressedWT <- rownames(statSigExpt1[order(statSigExpt1$log2FoldChange, decreasing = F)[1:sum(statSigExpt1$log2FoldChange < -0.5, na.rm=T)], ])

## Edit 20180220 I'll also make three list objects which house these result matrices
DSq2_resultsTableExpt1 <- list(as.matrix(statSigExpt1[order(statSigExpt1$log2FoldChange, decreasing = F)[1:sum(statSigExpt1$log2FoldChange < -1, na.rm=T)], ]), as.matrix(statSigExpt1[order(statSigExpt1$log2FoldChange, decreasing = TRUE)[1:sum(statSigExpt1$log2FoldChange > 1, na.rm=T)], ]))
# Names: change "Down IN B" to "OverexpressedWT"/OverExpressedNR - which can be checked using sampleTableExpt1, and can be easily renamed by copying the names from the names(EdgeDEgenes_Expt1) object which I've made below:
names(DSq2_resultsTableExpt1) <- names(EdgeDEgenes_Expt1)

DSq2_resultsTableExpt2 <- list(as.matrix(statSigExpt2[order(statSigExpt2$log2FoldChange, decreasing = F)[1:sum(statSigExpt2$log2FoldChange < -1, na.rm=T)], ]), as.matrix(statSigExpt2[order(statSigExpt2$log2FoldChange, decreasing = TRUE)[1:sum(statSigExpt2$log2FoldChange > 1, na.rm=T)], ]))  
names(DSq2_resultsTableExpt2) <- names(EdgeDEgenes_Expt2)

DSq2_resultsTableExpt3 <- list(as.matrix(statSigExpt3[order(statSigExpt3$log2FoldChange, decreasing = F)[1:sum(statSigExpt3$log2FoldChange < -1, na.rm=T)], ]), as.matrix(statSigExpt3[order(statSigExpt3$log2FoldChange, decreasing = TRUE)[1:sum(statSigExpt3$log2FoldChange > 1, na.rm=T)], ]))  
names(DSq2_resultsTableExpt3) <- names(EdgeDEgenes_Expt3)

save(DSq2_resultsTableExpt1,file='DSq2_resultsTableExpt1.RData')
save(DSq2_resultsTableExpt2,file='DSq2_resultsTableExpt2.RData')
save(DSq2_resultsTableExpt3,file='DSq2_resultsTableExpt3.RData')
save(Expt1lowStrin_OverExpressedNR,file='Expt1lowStrin_OverExpressedNR.RData')
save(Expt1lowStrin_OverExpressedWT,file='Expt1lowStrin_OverExpressedWT.RData')

```

Convert these up and downregulated gene names to something that can go through PANTHER
```{r}
XenbaseGenepageToGeneIdMapping <- read.delim("~/rnaSeqData/geneConverters/XenbaseGenepageToGeneIdMapping.txt", header=FALSE)
# readme:
##File contains the mapping of XB-GENEPAGE IDs, the primary identifier for a genepage to XB-GENE IDs for individual tropicalis and laevis genes.
##The file is tab-delimited and has the following format:
    #Xenbase genepage ID
    #Xenbase Gene IDs

# The ID in these next vectors is a sort of 'basal gene name', from which the names given to the transcript originates. e.g., the first gene in the following file is "nog", but prior to conversion is "nog.L" (which I know refers to the "long" chromosome in the X.laevis genome). This ".L" prevents any mapping
expt1_geneConversionUp <- as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(Expt1_upregInB,NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2]))
expt1_geneConversionDown <- as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(Expt1_downregInB,NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2]))

expt2_geneConversionUp <- as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(Expt2_upregInB,NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2]))
expt2_geneConversionDown <- as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(Expt2_downregInB,NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2]))

expt3_geneConversionUp <- as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(Expt3_upregInB,NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2]))
expt3_geneConversionDown <- as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(Expt3_downregInB,NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2]))

expt1lowStrin_geneConversionWT <- as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(Expt1lowStrin_OverExpressedWT,NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2]))
expt1lowStrin_geneConversionNR <- as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(Expt1lowStrin_OverExpressedNR,NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2]))

length(expt1_geneConversionDown)
length(expt1_geneConversionUp)
length(expt2_geneConversionDown)
length(expt2_geneConversionUp)
length(expt3_geneConversionDown)
length(expt3_geneConversionUp)


# The vectors above could essentially be used, but the conversion below filters for only genes with a human Ortholog - it tends to cut out all locXXXX transcripts. As far as I've been able to see with my eye, this is all that it does. 
XenbaseGeneHumanOrthologMapping <- read.delim("~/rnaSeqData/geneConverters/XenbaseGeneHumanOrthologMapping.txt", header=FALSE)

DownregGenes_expt1 <- as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(expt1_geneConversionDown,as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3])
length(DownregGenes_expt1)
length(expt1_geneConversionDown)

UpregGenes_expt1 <- as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(expt1_geneConversionUp,as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3])
length(UpregGenes_expt1)

DownregGenes_expt2 <- as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(expt2_geneConversionDown,as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3])
UpregGenes_expt2 <- as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(expt2_geneConversionUp,as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3])

DownregGenes_expt3 <- as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(expt3_geneConversionDown,as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3])
UpregGenes_expt3 <- as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(expt3_geneConversionUp,as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3])

Expt1lowStrin_NRoverExpressed <-as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(expt1lowStrin_geneConversionNR,as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3])
Expt1lowStrin_WToverExpressed <-as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(expt1lowStrin_geneConversionWT,as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3])
Expt1lowStrin_NRoverExpressed

write.csv(Expt1lowStrin_NRoverExpressed,file="DSq2DE_Expt1LowStrin_OverExpressedNR.csv")
write.csv(Expt1lowStrin_WToverExpressed,file="DSq2DE_Expt1LowStrin_OverExpressedWT.csv")

# Tidy up by removing the large objects involve in these conversions:
rm(NcbiMrnaXenbaseGene_laevis)
rm(XenbaseGeneHumanOrthologMapping)
rm(XenbaseGenepageToGeneIdMapping)





#### Put these items into a list which can be loaded as a single object in future 
#(and is probably how I should have been working with these objects)
# =====================================================================================
DESeqDEGenesExpt1 <- list(as.vector(DownregGenes_expt1),as.vector(UpregGenes_expt1))
# Find appropriate names that won't need a reference:
# Remember that "Down" and "Up" are relative to B (i.e., Down in B)
sampleTableExpt1 # While I had previously written these as down regulated in B (Non-regenerating) and Upregulated in B, I will now switch that to read Up in Regen (WT) and Up in NR
names(DESeqDEGenesExpt1) <- c("OverExpressedWT","OverExpressedNR")
DESeqDEGenesExpt1

DESeqDEGenesExpt2 <- list(as.vector(DownregGenes_expt2),as.vector(UpregGenes_expt2))
sampleTableExpt2
names(DESeqDEGenesExpt2) <- c("OverExpressedTbud","OverExpressedPNT")

DESeqDEGenesExpt3 <- list(as.vector(DownregGenes_expt3),as.vector(UpregGenes_expt3))
sampleTableExpt3
names(DESeqDEGenesExpt3) <- c("OverExpressedConCap","OverExpressedInjCap")

setwd("~/Documents/RnaR/rnaSeqRegen/Objects")
save(DESeqDEGenesExpt1,file='DESeqDEGenesExpt1.RData')
save(DESeqDEGenesExpt2,file='DESeqDEGenesExpt2.RData')
save(DESeqDEGenesExpt3,file='DESeqDEGenesExpt3.RData')

getwd()
# Will manually transfer to DESeq2_diffExpressedGenes
write.csv(DESeqDEGenesExpt1[[1]],file=paste0("DSq2DE_Expt1",names(DESeqDEGenesExpt1)[1],".csv"))
write.csv(DESeqDEGenesExpt1[[2]],file=paste0("DSq2DE_Expt1",names(DESeqDEGenesExpt1)[2],".csv"))

write.csv(DESeqDEGenesExpt2[[1]],file=paste0("DSq2DE_Expt2",names(DESeqDEGenesExpt2)[1],".csv"))
write.csv(DESeqDEGenesExpt2[[2]],file=paste0("DSq2DE_Expt2",names(DESeqDEGenesExpt2)[2],".csv"))

write.csv(DESeqDEGenesExpt3[[1]],file=paste0("DSq2DE_Expt3",names(DESeqDEGenesExpt3)[1],".csv"))
write.csv(DESeqDEGenesExpt3[[2]],file=paste0("DSq2DE_Expt3",names(DESeqDEGenesExpt3)[2],".csv"))
```



20180219
==========================================================================================================================================================

Two jobs: 
1. Use EdgeR to generate a new list of diff expressed genes. Can use the DE expt overlap between DESEq2 and EdgeR for a more robust analysis. 
2. Overlap EdgeR/DESeq2
  2.1 Might need to adjust the Expt1 parameters at this point, since Expt1 had very few DE genes - C states this may be due to the noisy nature of the experimental comparison. 
  


```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("BiocUpgrade")

biocLite("edgeR")
library("edgeR")
# Consulting the edgeR user guide: https://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf
# Two modules incorporated: classic (exact statistical models for multiple groups) and glm (generalised linear models). Both use empirical Bayes to estimate gene-specific biological variation
### EdgeR has a specific citation request, asking that citations are done (in addition to EdgeR) for the specific commands - i.e., it lists a bunch of papers with the various commands they have derived and asks for specific citation of these papers. 

# Within glm are two sub-modules: likelihood ratio test and quasi-likelihood F-test. Guide recommends the quasi-likelihood F-test for any bulk RNA-seq experiment, which uses a stricter error rate which can account for uncertainty in dispersion estimate (??). 
```

Reading counts from a file:
```{r}
# The formula: y <- DGEList(counts=x)
```
At this point, I assumed expt1TXI (search above) was the object I'm after. It contains the 'count' info. 
To re-familiarise myself with the TXI object I looked at the tximport vignette (https://bioconductor.org/packages/3.7/bioc/vignettes/tximport/inst/doc/tximport.html#edger)
Found this quick walkthrough:

cts <- txi$counts
normMat <- txi$length
normMat <- normMat/exp(rowMeans(log(normMat)))
library(edgeR)
o <- log(calcNormFactors(cts/normMat)) + log(colSums(cts/normMat))
y <- DGEList(cts)
y$offset <- t(t(log(normMat)) + o)
y is now ready for estimate dispersion functions see edgeR User's Guide

```{r}
expt1TXI <- tximport(expt1Files,type = "salmon", tx2gene = transcriptToGene)
expt2TXI <- tximport(expt2Files,type = "salmon", tx2gene = transcriptToGene)
expt3TXI <- tximport(expt3Files,type = "salmon", tx2gene = transcriptToGene)
length(expt1TXI)
names(expt1TXI)
# abundance, counts, length, countsFromAbundance

# Using the above framework:
cts <- expt1TXI$counts
normMat <- expt1TXI$length
normMat <- normMat/exp(rowMeans(log(normMat)))
library(edgeR)
o <- log(calcNormFactors(cts/normMat)) + log(colSums(cts/normMat))
y <- DGEList(cts)
y$offset <- t(t(log(normMat)) + o)
edgeR_Y_expt1 <- y

cts <- expt2TXI$counts
normMat <- expt2TXI$length
normMat <- normMat/exp(rowMeans(log(normMat)))
library(edgeR)
o <- log(calcNormFactors(cts/normMat)) + log(colSums(cts/normMat))
y <- DGEList(cts)
y$offset <- t(t(log(normMat)) + o)
edgeR_Y_expt2 <- y

cts <- expt3TXI$counts
normMat <- expt3TXI$length
normMat <- normMat/exp(rowMeans(log(normMat)))
library(edgeR)
o <- log(calcNormFactors(cts/normMat)) + log(colSums(cts/normMat))
y <- DGEList(cts)
y$offset <- t(t(log(normMat)) + o)
edgeR_Y_expt3 <- y

# Tidy up
rm(y)
rm(cts)
rm(normMat)
rm(o)

edgeR_Y_expt1$samples
# Lacks info about which group they are in
edgeR_Y_expt1$samples$group <- c(rep("1",3),rep("2",3))
edgeR_Y_expt1$samples
edgeR_Y_expt2$samples$group <- c(rep("1",3),rep("2",3))
edgeR_Y_expt3$samples$group <- c(rep("1",3),rep("2",3))

```


RNA-seq provides a measure of the relative abundance of each gene in each RNA sample, but does
not provide any measure of the total RNA output on a per-cell basis.  This commonly becomes
important  when  a  small  number  of  genes  are  very  highly  expressed  in  one  sample,  but  not  in
another.  The highly expressed genes can consume a substantial proportion of the total library size,
causing the remaining genes to be under-sampled in that sample.  Unless this
RNA  composition
effect is adjusted for, the remaining genes may falsely appear to be down-regulated in that sample
[30].
The
calcNormFactors
function normalizes for RNA composition by finding a set of scaling factors
for the library sizes that minimize the log-fold changes between the samples for most genes.  The
default method for computing these scale factors uses a trimmed mean of M-values (TMM) between
each pair of samples [30].  We call the product of the original library size and the scaling factor the
effective  library  size
.  The effective library size replaces the original library size in all downsteam
analyses.
TMM  is  the  recommended  for  most  RNA-Seq  data  where  the  majority  (more  than  half)  of
the genes are believed not differentially expressed between any pair of the samples.  The following
commands perform the TMM normalization and display the normalization factors.
# y <- calcNormFactors(y)
# y$samples
```{r}
edgeR_Y_expt1 <- calcNormFactors(edgeR_Y_expt1)
edgeR_Y_expt1$samples

edgeR_Y_expt2 <- calcNormFactors(edgeR_Y_expt2)
edgeR_Y_expt2$samples

edgeR_Y_expt3 <- calcNormFactors(edgeR_Y_expt3)
edgeR_Y_expt3$samples
```



Pairwise comparison between two groups
----------------------------------------
Estimating dispersions:
A glm (generalised linear model) estimates probabiblity distributions according to their mean-variance relationship. This first involves calculating "dispersion". 

```{r}
# y <- estimateDisp(y, design)
# To create the design object:

designExpt1 <- model.matrix(~0+group, data=edgeR_Y_expt1$samples)
colnames(designExpt1) <- c("A","B")
designExpt2 <- model.matrix(~0+group, data=edgeR_Y_expt2$samples)
colnames(designExpt1) <- c("A","B")
designExpt3 <- model.matrix(~0+group, data=edgeR_Y_expt3$samples)
colnames(designExpt1) <- c("A","B")

edgeR_dispersion_expt1 <- estimateDisp(edgeR_Y_expt1,designExpt1)
edgeR_dispersion_expt2 <- estimateDisp(edgeR_Y_expt2,designExpt2)
edgeR_dispersion_expt3 <- estimateDisp(edgeR_Y_expt3,designExpt3)
```

# Now ready to test for DE genes using a linear model approach:
```{r}
fitExpt1 <- glmQLFit(edgeR_dispersion_expt1, designExpt1)
fitExpt2 <- glmQLFit(edgeR_dispersion_expt2, designExpt2)
fitExpt3 <- glmQLFit(edgeR_dispersion_expt3, designExpt3)

BvsA <- makeContrasts(B-A, levels=designExpt1)
# Which is the same for experiments 2 and 3
BvsA

glmResultExpt1 <- glmQLFTest(fitExpt1, contrast=BvsA)
glmResultExpt2 <- glmQLFTest(fitExpt2, contrast=BvsA)
glmResultExpt3 <- glmQLFTest(fitExpt3, contrast=BvsA)

topTags(glmResultExpt1)
?topTags
topTags(glmResultExpt1,adjust.method = "BH", sort.by = "logFC")
# there should be a way to retrieve all tags with p values < 0.05:
glmExpt1_p05 <- topTags(glmResultExpt1,p.value = 0.05)
dim(glmExpt1_p05)
# Doesn't work, only returns the top 10. 

topTags(glmResultExpt1,n=100)
# This can retrieve as many as specified:
topTags(glmResultExpt1)
glmExpt1_p05 <- topTags(glmResultExpt1,n=50000,adjust.method="BH",p.value = 0.05) # where n is any number larger than the number of genes with P < 0.05
glmExpt2_p05 <- topTags(glmResultExpt2,n=50000,adjust.method="BH",p.value = 0.05)
glmExpt3_p05 <- topTags(glmResultExpt3,n=50000,adjust.method="BH",p.value = 0.05)

# Pull out only genes which are also biologically (log2FC)
## The help page for glmLRT says: "The data frame table contains the following columns: logFC: log2-fold change of expression between conditions being tested." so I know I'm working oin log2FC. (https://support.bioconductor.org/p/50273/)
colnames(glmExpt1_p05)
edgerSigResultsUpRegExpt1 <- glmExpt1_p05[order(glmExpt1_p05$table$logFC,decreasing=T)[1:sum(sort(glmExpt1_p05$table$logFC,decreasing=T) >1,na.rm=T)],]
edgerSigResultsUpRegExpt1[1:20,]
edgerSigResultsDownRegExpt1 <- glmExpt1_p05[order(glmExpt1_p05$table$logFC,decreasing=F)[1:sum(sort(glmExpt1_p05$table$logFC,decreasing=F) < -1,na.rm=T)],]
edgerSigResultsDownRegExpt1[1:20,]

grep("nog",rownames(edgerSigResultsDownRegExpt1))
grep("nog",rownames(edgerSigResultsUpRegExpt1))

dim(edgerSigResultsDownRegExpt1)
dim(edgerSigResultsUpRegExpt1)

# Expt2
edgerSigResultsUpRegExpt2 <- glmExpt2_p05[order(glmExpt2_p05$table$logFC,decreasing=T)[1:sum(sort(glmExpt2_p05$table$logFC,decreasing=T) >1,na.rm=T)],]
edgerSigResultsUpRegExpt2[1:20,]
edgerSigResultsDownRegExpt2 <- glmExpt2_p05[order(glmExpt2_p05$table$logFC,decreasing=F)[1:sum(sort(glmExpt2_p05$table$logFC,decreasing=F) < -1,na.rm=T)],]
edgerSigResultsDownRegExpt2[1:20,]
dim(edgerSigResultsDownRegExpt2)
dim(edgerSigResultsUpRegExpt2)

# Expt3
edgerSigResultsUpRegExpt3 <- glmExpt3_p05[order(glmExpt3_p05$table$logFC,decreasing=T)[1:sum(sort(glmExpt3_p05$table$logFC,decreasing=T) >1,na.rm=T)],]
edgerSigResultsUpRegExpt3[1:20,]
edgerSigResultsDownRegExpt3 <- glmExpt3_p05[order(glmExpt3_p05$table$logFC,decreasing=F)[1:sum(sort(glmExpt3_p05$table$logFC,decreasing=F) < -1,na.rm=T)],]
edgerSigResultsDownRegExpt3[1:20,]
dim(edgerSigResultsDownRegExpt3)
dim(edgerSigResultsUpRegExpt3)
```


Noticably, there are many more DE genes in the EdgeR analysis compared to the DESeq2 analysis. 
- I have not yet been able to figure out if EdgeR moderates genes on the basis of low-count > high variance. This did appear to make a substantial difference when implemented in the DESeq2 package. 
- There are no doubt other differences in the default settings (https://mikelove.wordpress.com/2016/09/28/deseq2-or-edger/)

I can now calculate the overlap between EdgeR and DESeq2 - it will be important to look at the overlap from the point of view of agreement (i.e., the two support each other's calls) as well as 'final list' (DE genes are only those contained in both lists). 
  - The second of these could be problematic, since I know that DESEq2 for some reason had few DE genes in Expt1. 
  
Saving a copy of the environment
```{r}
# In Documents/RnaR/rnaSeqRegen/
save.image("~/Documents/RnaR/rnaSeqRegen/DiffExpressedGenes_environment.RData")
```

  
  
20190220
==============================================================================================================================================================================
Today:
Retrieve the geneSymbol for Edger DE genes (currently they are in the nog.l format)

Overlap between EdgeR and DSq2. 
Explore the parameters in DSq2 which could lead to a larger number of DE genes in Expt1. 
  - This will run as a secondary figure/subsection - not a replacement.

```{r}
# ========================================
## Save three list objects which have both up and down reg genes for each expt
# remember to save the names of the lists as OverexpressedNR OverexpressedWT, which will be easy to interpret without loading and referring to a sample table in the future. 
EdgeR_resultsTableExpt1 <- list(edgerSigResultsDownRegExpt1,edgerSigResultsUpRegExpt1)
designExpt1
sampleTableExpt1
names(EdgeR_resultsTableExpt1) <- c("OverExpressedWT","OverExpressedNR")

EdgeR_resultsTableExpt2 <- list(edgerSigResultsDownRegExpt2,edgerSigResultsUpRegExpt2)
designExpt2
sampleTableExpt2
names(EdgeR_resultsTableExpt2) <- c("OverExpressedTbud","OverExpressedPNT")

EdgeR_resultsTableExpt3 <- list(edgerSigResultsDownRegExpt3,edgerSigResultsUpRegExpt3)
designExpt3
sampleTableExpt3
names(EdgeR_resultsTableExpt3) <- c("OverExpressedConCap","OverExpressedInjCap")

# Save objects:
save(EdgeR_resultsTableExpt1,file="EdgeR_resultsTableExpt1.RData")
save(EdgeR_resultsTableExpt2,file="EdgeR_resultsTableExpt2.RData")
save(EdgeR_resultsTableExpt3,file="EdgeR_resultsTableExpt3.RData")
```


Retrieve gene symbols:
```{r}
XenbaseGenepageToGeneIdMapping <- read.delim("~/rnaSeqData/geneConverters/XenbaseGenepageToGeneIdMapping.txt", header=FALSE)
# The ID in these next vectors is a sort of 'basal gene name', from which the names given to the transcript originates. e.g., the first gene in the following file is "nog", but prior to conversion is "nog.L" (which I know refers to the "long" chromosome in the X.laevis genome). This ".L" prevents any mapping
NcbiMrnaXenbaseGene_laevis <- read.delim("~/rnaSeqData/geneConverters/NcbiMrnaXenbaseGene_laevis.txt", header=FALSE)

EdgeExpt1 <- lapply(1:length(EdgeR_resultsTableExpt1), function(x) as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(rownames(EdgeR_resultsTableExpt1[[x]]),NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2])))
# Remember that not all genes will match, will therefore need to trim the table based on how many do map
EdgeExpt2 <- lapply(1:length(EdgeR_resultsTableExpt2), function(x) as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(rownames(EdgeR_resultsTableExpt2[[x]]),NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2])))
EdgeExpt3 <- lapply(1:length(EdgeR_resultsTableExpt3), function(x) as.vector(na.omit(XenbaseGenepageToGeneIdMapping[match(NcbiMrnaXenbaseGene_laevis[as.vector(na.omit(match(rownames(EdgeR_resultsTableExpt3[[x]]),NcbiMrnaXenbaseGene_laevis[,4]))),3],XenbaseGenepageToGeneIdMapping[,5]),2])))

names(EdgeExpt1) <- names(EdgeR_resultsTableExpt1)
names(EdgeExpt2) <- names(EdgeR_resultsTableExpt2)
names(EdgeExpt3) <- names(EdgeR_resultsTableExpt3)

# The vectors above could essentially be used, but the conversion below filters for only genes with a human Ortholog - it tends to cut out all locXXXX transcripts. As far as I've been able to see with my eye, this is all that it does. 

XenbaseGeneHumanOrthologMapping <- read.delim("~/rnaSeqData/geneConverters/XenbaseGeneHumanOrthologMapping.txt", header=FALSE)
EdgeDEgenes_Expt1 <- lapply(1:length(EdgeExpt1), function(x) as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(EdgeExpt1[[x]],as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3]))
names(EdgeDEgenes_Expt1) <- names(EdgeExpt1)

EdgeDEgenes_Expt2 <- lapply(1:length(EdgeExpt2), function(x) as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(EdgeExpt2[[x]],as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3]))
names(EdgeDEgenes_Expt2) <- names(EdgeExpt2)

EdgeDEgenes_Expt3 <- lapply(1:length(EdgeExpt3), function(x) as.vector(XenbaseGeneHumanOrthologMapping[as.vector(na.omit(match(EdgeExpt3[[x]],as.vector(XenbaseGeneHumanOrthologMapping[,3])))),3]))
names(EdgeDEgenes_Expt3) <- names(EdgeExpt3)

EdgeDEgenes_Expt1

# Save objects:
save(EdgeDEgenes_Expt1,file='EdgeDEgenes_Expt1.RData')
save(EdgeDEgenes_Expt2,file='EdgeDEgenes_Expt2.RData')
save(EdgeDEgenes_Expt3,file='EdgeDEgenes_Expt3.RData')

getwd()
# Will manually transfer to EdgeR_diffExpressedGenes
write.csv(EdgeDEgenes_Expt1[[1]],file=paste0("EdgeDE_Expt1",names(EdgeDEgenes_Expt1)[1],".csv"))
write.csv(EdgeDEgenes_Expt1[[2]],file=paste0("EdgeDE_Expt1",names(EdgeDEgenes_Expt1)[2],".csv"))

write.csv(EdgeDEgenes_Expt2[[1]],file=paste0("EdgeDE_Expt2",names(EdgeDEgenes_Expt2)[1],".csv"))
write.csv(EdgeDEgenes_Expt2[[2]],file=paste0("EdgeDE_Expt2",names(EdgeDEgenes_Expt2)[2],".csv"))

write.csv(EdgeDEgenes_Expt3[[1]],file=paste0("EdgeDE_Expt3",names(EdgeDEgenes_Expt3)[1],".csv"))
write.csv(EdgeDEgenes_Expt3[[2]],file=paste0("EdgeDE_Expt3",names(EdgeDEgenes_Expt3)[2],".csv"))
```


Calculate overlap between DE genes in EdgeR and DSq2
```{r}
## Expt1
length(rownames(EdgeR_resultsTableExpt1[[1]])) # 84
length(rownames(DSq2_resultsTableExpt1[[1]])) #28
length(na.omit(match(rownames(DSq2_resultsTableExpt1[[1]]),rownames(EdgeR_resultsTableExpt1[[1]]))))
# 28/28 = 100%, 28/84 = 33%
length(rownames(DSq2_resultsTableExpt1[[2]])) #49
length(rownames(EdgeR_resultsTableExpt1[[2]])) #234
length(na.omit(match(rownames(DSq2_resultsTableExpt1[[2]]),rownames(EdgeR_resultsTableExpt1[[2]]))))
# 49/49 = 100%, 49/234=21%

## Expt2
length(rownames(DSq2_resultsTableExpt2[[1]])) #178
length(rownames(EdgeR_resultsTableExpt2[[1]]))  # 237
length(na.omit(match(rownames(DSq2_resultsTableExpt2[[1]]),rownames(EdgeR_resultsTableExpt2[[1]]))))
# 177/178 = 99%, 177/237 = 75%
length(rownames(DSq2_resultsTableExpt2[[2]])) #890
length(rownames(EdgeR_resultsTableExpt2[[2]]))  # 1020
length(na.omit(match(rownames(DSq2_resultsTableExpt2[[2]]),rownames(EdgeR_resultsTableExpt2[[2]]))))
# 887/890 = 99%, 887/1020 = 87%

## Expt3
length(rownames(DSq2_resultsTableExpt3[[1]])) #1588
length(rownames(EdgeR_resultsTableExpt3[[1]]))  # 1636
length(na.omit(match(rownames(DSq2_resultsTableExpt3[[1]]),rownames(EdgeR_resultsTableExpt3[[1]]))))
# 1577/1588 = 99%, 1577/1636 = 96%
length(rownames(DSq2_resultsTableExpt3[[2]])) #1753
length(rownames(EdgeR_resultsTableExpt3[[2]]))  # 1845
length(na.omit(match(rownames(DSq2_resultsTableExpt3[[2]]),rownames(EdgeR_resultsTableExpt3[[2]]))))
# 1738/1753 = 99%, 1738/1845 = 94%
```

These are all >99% agreement. This gives me good confidence in the analyses I've done so far, there is nothing so drastically wrong as to throw off the entire analysis. 

I had intended to overlap DSq2 and EdgeR gene lists, but since the overlap is so high, there is very little to be gained from this that I don't get by just using the DSq2 lists. 

Use the results table rownames to find out how many genes are found in common between the experiments
```{r}
dim(DSq2_resultsTableExpt2[[1]]) # 178 genes
dim(DSq2_resultsTableExpt2[[2]]) # 890 genes
dim(DSq2_resultsTableExpt3[[1]]) # 1588 genes
dim(DSq2_resultsTableExpt3[[2]]) # 1753 genes
length(na.omit(match(rownames(DSq2_resultsTableExpt2[[1]]),rownames(DSq2_resultsTableExpt3[[1]])))) # Overlap is 8
length(na.omit(match(rownames(DSq2_resultsTableExpt2[[1]]),rownames(DSq2_resultsTableExpt3[[2]])))) # Overlap is 115
length(na.omit(match(rownames(DSq2_resultsTableExpt2[[2]]),rownames(DSq2_resultsTableExpt3[[2]])))) # Overlap is 308
length(na.omit(match(rownames(DSq2_resultsTableExpt2[[2]]),rownames(DSq2_resultsTableExpt3[[1]])))) # Overlap is 156
```



Now I will use the DEgenes list objects (as opposed to the resultsTable objects) which have the proper geneSymbol and create overlap lists
Expt1/2/3 overlaps
```{r}
length(na.omit(match(DESeqDEGenesExpt1[[1]], DESeqDEGenesExpt2[[1]])))
# 5
length(na.omit(match(DESeqDEGenesExpt1[[1]], DESeqDEGenesExpt2[[2]])))
# 3
length(na.omit(match(DESeqDEGenesExpt1[[1]], DESeqDEGenesExpt3[[1]])))
# 0
length(na.omit(match(DESeqDEGenesExpt1[[1]], DESeqDEGenesExpt3[[2]])))
# 12

length(na.omit(match(DESeqDEGenesExpt1[[2]], DESeqDEGenesExpt2[[1]])))
# 2
length(na.omit(match(DESeqDEGenesExpt1[[2]], DESeqDEGenesExpt2[[2]])))
# 24
length(na.omit(match(DESeqDEGenesExpt1[[2]], DESeqDEGenesExpt3[[1]])))
# 5
length(na.omit(match(DESeqDEGenesExpt1[[2]], DESeqDEGenesExpt3[[2]])))
# 13


length(na.omit(match(DESeqDEGenesExpt2[[1]], DESeqDEGenesExpt2[[1]])))
## same list
length(na.omit(match(DESeqDEGenesExpt2[[1]], DESeqDEGenesExpt2[[2]])))
## Mutually exclusive list
length(na.omit(match(DESeqDEGenesExpt2[[1]], DESeqDEGenesExpt3[[1]])))
# 7
length(na.omit(match(DESeqDEGenesExpt2[[1]], DESeqDEGenesExpt3[[2]])))
# 78

length(na.omit(match(DESeqDEGenesExpt2[[2]], DESeqDEGenesExpt2[[1]])))
## Mutually exclusive list
length(na.omit(match(DESeqDEGenesExpt2[[2]], DESeqDEGenesExpt2[[2]])))
## same list
length(na.omit(match(DESeqDEGenesExpt2[[2]], DESeqDEGenesExpt3[[1]])))
# 129
length(na.omit(match(DESeqDEGenesExpt2[[2]], DESeqDEGenesExpt3[[2]])))
# 239

## Interesting
# Experiment3_2 (OverExpressedInjCap) shares 78 genes with Expt2_1 (overExpressedTBud) and 239 genes with Expt2_2 (overExpressedPNT), It also shares 13 genes with Expt1_2 (overexpressed in NR) (n=38 genes) which is a 34% overlap. 
# Expt2_2 and 3_2 have a high overlap (n=239) but there are also a lot of genes in these groups (n=709 and 1344 respectively) > overlap percentage is 239/709 = 34%. 

write.csv(DESeqDEGenesExpt3[[2]][as.vector(na.omit(match(DESeqDEGenesExpt2[[2]], DESeqDEGenesExpt3[[2]])))],file='Genes overexpressed in PNT and InjCap.csv')
write.csv(DESeqDEGenesExpt3[[1]][as.vector(na.omit(match(DESeqDEGenesExpt2[[2]], DESeqDEGenesExpt3[[1]])))],file='Genes overexpressed in PNT and ConCap.csv')
# This is kind of strange. Overexpressed in PNT relative to the tailbud are found in both InjCap and ConCap
## Suggests that overexpressed in PNT (n=709) is really a heterogenous group of genes with two separate functions. It's got a larger subgroup which is overexpressed in InjCap (n=239) and a smaller subgroup of genes overexpressed in ConCap (n=129).  


write.csv(DESeqDEGenesExpt3[[2]][as.vector(na.omit(match(DESeqDEGenesExpt2[[1]], DESeqDEGenesExpt3[[2]])))],file='Genes overexpressed in InjCap and Tbud.csv')
write.csv(DESeqDEGenesExpt3[[1]][as.vector(na.omit(match(DESeqDEGenesExpt2[[1]], DESeqDEGenesExpt3[[1]])))],file='Genes overexpressed in Tbud and ConCap.csv')

write.csv(DESeqDEGenesExpt3[[2]][as.vector(na.omit(match(DESeqDEGenesExpt1[[1]], DESeqDEGenesExpt3[[2]])))],file='Genes overexpressed in InjCap and WT.csv')

# After talking with C, I realise that PNT is a sort of 'noise/maintenance background' set (i.e., it's not regenerating)
# - in fact, some of these genes may need to be turned off in order to facilitate regeneration.
## I'll now make a list of e.g., Injcap minus PNT

write.csv(DESeqDEGenesExpt3[[2]][-as.vector(na.omit(match(DESeqDEGenesExpt2[[2]],DESeqDEGenesExpt3[[2]])))],file='Overexpressed in InjCap NOT in PNT.csv')
write.csv(DESeqDEGenesExpt3[[1]][-as.vector(na.omit(match(DESeqDEGenesExpt2[[2]],DESeqDEGenesExpt3[[1]])))],file='Overexpressed in ConCap NOT in PNT.csv')

write.csv(DESeqDEGenesExpt2[[1]][-as.vector(na.omit(match(DESeqDEGenesExpt3[[1]],DESeqDEGenesExpt2[[1]])))],file='Overexpressed in Tbud NOT in ConCap.csv')
length(DESeqDEGenesExpt2[[1]][-as.vector(na.omit(match(DESeqDEGenesExpt3[[1]],DESeqDEGenesExpt2[[1]])))]) # 118
length(DESeqDEGenesExpt2[[1]])
```


## Edit 20180313
I want to repeat this overlap data to take into account the new low stringency Expt1 gene lists.
```{r}
Expt1lowStrin_NRoverExpressed
Expt1lowStrin_WToverExpressed

names(DESeqDEGenesExpt2)
# [1] "OverExpressedTbud" "OverExpressedPNT"
names(DESeqDEGenesExpt3)
# [1] "OverExpressedConCap" "OverExpressedInjCap"

length(na.omit(match(Expt1lowStrin_WToverExpressed,DESeqDEGenesExpt2[[1]])))    # 18 genes overlap between tbud and WT
length(na.omit(match(Expt1lowStrin_WToverExpressed,DESeqDEGenesExpt2[[2]])))    # 16 genes overlap between PNT and WT
length(na.omit(match(Expt1lowStrin_WToverExpressed,DESeqDEGenesExpt3[[1]])))    # 6 overlap between conCap and WT
length(na.omit(match(Expt1lowStrin_WToverExpressed,DESeqDEGenesExpt3[[2]])))    # 139 genes overlap between WT and InjCap
length(DESeqDEGenesExpt3[[2]]) ## WT is 201 genes long, InjCap is 1344, 139 overlap is 69% and 10% respectively. 

length(na.omit(match(Expt1lowStrin_NRoverExpressed,DESeqDEGenesExpt2[[1]])))    # 15 genes overlap between tbud and NR
length(na.omit(match(Expt1lowStrin_NRoverExpressed,DESeqDEGenesExpt2[[2]])))    # 153 genes overlap between PNT and NR
length(na.omit(match(Expt1lowStrin_NRoverExpressed,DESeqDEGenesExpt3[[1]])))    # 71 overlap between conCap and NR
length(na.omit(match(Expt1lowStrin_NRoverExpressed,DESeqDEGenesExpt3[[2]])))    # 76 genes overlap between InjCap and NR


## InjCap shares 139 genes with WT, but also shares 76 genes with NR. 

lowStrinWT_plus_Tbud <- Expt1lowStrin_WToverExpressed[na.omit(match(DESeqDEGenesExpt2[[1]],Expt1lowStrin_WToverExpressed))]
lowStrinWT_plus_Pnt <- Expt1lowStrin_WToverExpressed[na.omit(match(DESeqDEGenesExpt2[[2]],Expt1lowStrin_WToverExpressed))]
lowStrinWT_plus_conCap <- Expt1lowStrin_WToverExpressed[na.omit(match(DESeqDEGenesExpt3[[1]],Expt1lowStrin_WToverExpressed))]
lowStrinWT_plus_injCap <- Expt1lowStrin_WToverExpressed[na.omit(match(DESeqDEGenesExpt3[[2]],Expt1lowStrin_WToverExpressed))]

lowStrinNR_plus_Tbud <- Expt1lowStrin_NRoverExpressed[na.omit(match(DESeqDEGenesExpt2[[1]],Expt1lowStrin_NRoverExpressed))]
lowStrinNR_plus_Pnt <- Expt1lowStrin_NRoverExpressed[na.omit(match(DESeqDEGenesExpt2[[2]],Expt1lowStrin_NRoverExpressed))]
lowStrinNR_plus_conCap <-Expt1lowStrin_NRoverExpressed[na.omit(match(DESeqDEGenesExpt3[[1]],Expt1lowStrin_NRoverExpressed))]
lowStrinNR_plus_injCap <- Expt1lowStrin_NRoverExpressed[na.omit(match(DESeqDEGenesExpt3[[2]],Expt1lowStrin_NRoverExpressed))]


write.csv(lowStrinWT_plus_Tbud,file='lowStrinWT_plus_Tbud.csv')
write.csv(lowStrinWT_plus_Pnt,file='lowStrinWT_plus_Pnt.csv')
write.csv(lowStrinWT_plus_conCap,file='lowStrinWT_plus_conCap.csv')
write.csv(lowStrinWT_plus_injCap,file='lowStrinWT_plus_injCap.csv')

write.csv(lowStrinNR_plus_Tbud,file='lowStrinNR_plus_Tbud.csv')
write.csv(lowStrinNR_plus_Pnt,file='lowStrinNR_plus_Pnt.csv')
write.csv(lowStrinNR_plus_conCap,file='lowStrinNR_plus_conCap.csv')
write.csv(lowStrinNR_plus_injCap,file='lowStrinNR_plus_injCap.csv')
```



Use GO to find out what each of the gene subsets are associated with. 






20180313
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Have now used PANTHER to retrieve gene ontology (Biol Slim) tables for each of the gene lists. 
Also generated panther Pathway gene lists (the list of DE genes which are actually found in an enriched pathway).
I want to load those into R and look at what overlaps and what does not. I'll also be looking for the genes which C has cherry-picked and are being cloned. 

I've just saved the environment, which contains all of the objects used to date, as diffExpressedGenes_environment_20180313. 
Clear the workspace, and load in the various pathway lists and C's cherry-picked genes (Cgenes). 

```{r}
injCapGeneList_hedgehog <- read.delim("~/Documents/RnaR/rnaSeqRegen/PANTHERDB/DSq2/injCap/injCap_pathwaygeneLists/injCapGeneList_hedgehog.txt", header=FALSE)

# But rather than read in a single file at a time, I want to read in all files in a directory.

# InjCap
getwd()
injCapDir <- "~/Documents/RnaR/rnaSeqRegen/PANTHERDB/DSq2/injCap/injCap_pathwaygeneLists/"
injCapFilename <- list.files(injCapDir)
injCapFilename

geneListColumnNames <- c("UniProt","GeneSymbol","DescriptorA","DescriptorB","DescriptorC","Unknown")
injCapPathwayGeneLists <- lapply(1:length(injCapFilename),function(x) read.delim(paste0(injCapDir,injCapFilename)[[x]],header=F,col.names=geneListColumnNames))
names(injCapPathwayGeneLists) <- injCapFilename
injCapPathwayGeneLists[1]
rm(injCapDir)
rm(injCapFilename)

# Ok: this works, but would then invovle copy pasting these 14 lines of code and replacing "InjCap" with "conCap/Tbud" etc., which I don't want to do. I can re-write the code to be less specific, and then only need to change the first (e.g., thePathwayDir) and last (final object name) names. 
# What I can do as an alternative is try and list all the file names in DSq2 and then write a loop to run through them all. But this would be frustrating to do. 

# Non-specific pathway
geneListColumnNames <- c("UniProt","GeneSymbol","DescriptorA","DescriptorB","DescriptorC","Unknown")

# Non-specific pathway
getwd()
Dir <- "~/Documents/RnaR/rnaSeqRegen/PANTHERDB/DSq2/injCap/injCap_pathwayGeneLists/"
Filename <- list.files(Dir)
Filename
PathwayGeneLists <- lapply(1:length(Filename),function(x) read.delim(paste0(Dir,Filename)[[x]],header=F,col.names=geneListColumnNames))
names(PathwayGeneLists) <- Filename
PathwayGeneLists[1]
injCapPathwayGeneLists <- PathwayGeneLists
rm(PathwayGeneLists)
rm(Dir)
rm(Filename)


Dir <- "~/Documents/RnaR/rnaSeqRegen/PANTHERDB/DSq2/conCap/conCap_pathwayGeneLists/"
Filename <- list.files(Dir)
Filename
PathwayGeneLists <- lapply(1:length(Filename),function(x) read.delim(paste0(Dir,Filename)[[x]],header=F,col.names=geneListColumnNames))
names(PathwayGeneLists) <- Filename
PathwayGeneLists[1]
conCapPathwayGeneLists <- PathwayGeneLists
rm(PathwayGeneLists)
rm(Dir)
rm(Filename)


Dir <- "~/Documents/RnaR/rnaSeqRegen/PANTHERDB/DSq2/NRlowStrin/NRlowStrin_pathwayGeneLists/"
Filename <- list.files(Dir)
Filename
PathwayGeneLists <- lapply(1:length(Filename),function(x) read.delim(paste0(Dir,Filename)[[x]],header=F,col.names=geneListColumnNames))
names(PathwayGeneLists) <- Filename
PathwayGeneLists[1]
NRlowStrinPathwayGeneLists <- PathwayGeneLists
rm(PathwayGeneLists)
rm(Dir)
rm(Filename)


Dir <- "~/Documents/RnaR/rnaSeqRegen/PANTHERDB/DSq2/pnt/pnt_pathwayGeneLists/"
Filename <- list.files(Dir)
Filename
PathwayGeneLists <- lapply(1:length(Filename),function(x) read.delim(paste0(Dir,Filename)[[x]],header=F,col.names=geneListColumnNames))
names(PathwayGeneLists) <- Filename
PathwayGeneLists[1]
pntPathwayGeneLists <- PathwayGeneLists
rm(PathwayGeneLists)
rm(Dir)
rm(Filename)


Dir <- "~/Documents/RnaR/rnaSeqRegen/PANTHERDB/DSq2/Tbud/Tbud_pathwayGeneLists/"
Filename <- list.files(Dir)
Filename
PathwayGeneLists <- lapply(1:length(Filename),function(x) read.delim(paste0(Dir,Filename)[[x]],header=F,col.names=geneListColumnNames))
names(PathwayGeneLists) <- Filename
PathwayGeneLists[1]
TbudPathwayGeneLists <- PathwayGeneLists
rm(PathwayGeneLists)
rm(Dir)
rm(Filename)


Dir <- "~/Documents/RnaR/rnaSeqRegen/PANTHERDB/DSq2/WTlowStrin/WTlowStrin_pathwayGeneLists/"
Filename <- list.files(Dir)
Filename
PathwayGeneLists <- lapply(1:length(Filename),function(x) read.delim(paste0(Dir,Filename)[[x]],header=F,col.names=geneListColumnNames))
names(PathwayGeneLists) <- Filename
PathwayGeneLists[1]
WTlowStrinPathwayGeneLists <- PathwayGeneLists
rm(PathwayGeneLists)
rm(Dir)
rm(Filename)

## now to add the list of Cgenes:
Cgenes <-c("inhba","phdgh","not","spry4","hoxa13","fgf20","crabp2","T","prss27","irrc15","sp5l","gli1","endou","apcdd","adamts18","sox10","endrb")
## I've added gli1, which is listed on the Cgenes cloning planner as being the same as sp5l
lapply(1:length(injCapPathwayGeneLists),function(x) length(na.omit(match(Cgenes,injCapPathwayGeneLists[[x]]$GeneSymbol))))

pantherPathwayGeneLists <- list(WTlowStrinPathwayGeneLists,NRlowStrinPathwayGeneLists,TbudPathwayGeneLists,pntPathwayGeneLists,conCapPathwayGeneLists,injCapPathwayGeneLists)

names(pantherPathwayGeneLists) <- c("WTlowStrinPathway","NRlowStrinPathway","TbudPathway","pntPathway","conCapPathway","injCapPathway")
unlist(lapply(1:length(pantherPathwayGeneLists[[5]]),function(x) length(na.omit(match(Cgenes,as.vector(pantherPathwayGeneLists[[5]][[x]]$GeneSymbol)))))) 
# TbudPathway[1] has 1 gene
TbudPathwayGeneLists[[1]]$GeneSymbol[na.omit(match(Cgenes,TbudPathwayGeneLists[[1]]$GeneSymbol))]

```























